src/trezor.rs

```diff
@@ -0,0 +1,185 @@
+        } else {
+            Err(Error::DeviceNotFound)
+        }
+    }
+}
+
+#[async_trait]
+impl HWI for TrezorClient {
+    fn device_kind(&self) -> crate::DeviceKind {
+        self.kind
+    }
+
+    async fn get_version(&self) -> Result<super::Version, Error> {
+        let client = self.client.borrow();
+        let f = client.features();
+        if let Some(f) = f {
+            let version = super::Version {
+                major: f.major_version(),
+                minor: f.minor_version(),
+                patch: f.patch_version(),
+                prerelease: None,
+            };
+            Ok(version)
+        } else {
+            return Err(Error::Device(String::from("No features found")));
+        }
+    }
+
+    async fn is_connected(&self) -> Result<(), Error> {
+        match self.client.borrow_mut().ping("PINGPING")? {
+            TrezorResponse::Ok(_) => Ok(()),
+            _ => Err(Error::DeviceDisconnected),
+        }
+    }
+    async fn get_master_fingerprint(&self) -> Result<Fingerprint, Error> {
+        let path = DerivationPath::default();
+        match self.client.borrow_mut().get_public_key(
+            &path,
+            trezor_client::InputScriptType::SPENDADDRESS,
+            bitcoin::Network::Bitcoin,
+            false,
+        ) {
+            Ok(TrezorResponse::Ok(key)) => {
+                let fp = key.fingerprint();
+                Ok(fp)
+            }
+            Ok(TrezorResponse::Failure(f)) => Err(Error::Device(f.to_string())),
+            Ok(result) => Err(Error::Device(result.to_string())),
+            Err(e) => Err(Error::Device(e.to_string())),
+        }
+    }
+
+    async fn get_extended_pubkey(&self, path: &DerivationPath) -> Result<ExtendedPubKey, Error> {
+        let path = DerivationPath::from_str(&path.to_string())
+            .map_err(|e| Error::Device(format!("{:#?}", e)))?;
+        match self.client.borrow_mut().get_public_key(
+            &path,
+            trezor_client::InputScriptType::SPENDADDRESS,
+            bitcoin::Network::Bitcoin,
+            false,
+        ) {
+            Ok(TrezorResponse::Ok(key)) => {
+                return Ok(key);
+            }
+            Ok(TrezorResponse::Failure(f)) => Err(Error::Device(f.to_string())),
+            Ok(result) => Err(Error::Device(result.to_string())),
+            Err(e) => Err(Error::Device(e.to_string())),
+        }
+    }
+    async fn register_wallet(&self, _name: &str, _policy: &str) -> Result<Option<[u8; 32]>, Error> {
+        return Err(Error::UnimplementedMethod);
+    }
+    async fn sign_tx(&self, tx: &mut Psbt) -> Result<(), Error> {
+        let mut client = self.client.borrow_mut();
+        let data = client.sign_tx(tx, self.network)?;
+        if let TrezorResponse::Ok(progress) = data {
+            if progress.has_signature() {
+                return Ok(());
+            }
+        }
+        return Err(Error::DeviceDidNotSign);
+    }
+}
+
+impl TrezorClient {
```

Comment:
You have two `impl TrezorClient` blocks, I understand that this one is for the simulator, but it is better in my opinion to merge them.

Reply:
Merged both of them. I didn't have any strong opinions either way.